### PR TITLE
installplan pkg added

### PR DESCRIPTION
**Pull request by elenagerman**

### CODE REVIEW

**Comment by kononovn** [comment]
Hi @elenagerman This pkg should be a part of ./pkg/olm package

File: pkg/olm/installplan.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Builder provides a struct for installplan object from the cluster and a installplan definition.
+type Builder struct {
+	// installplan definition, used to create the installplan object.
```
Comment (kononovn) [inline pkg/olm/installplan.go:18]: Minor: Please use capital letter. Installplan

File: pkg/olm/installplan.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+		glog.V(100).Infof("The nsname of the installplan is empty")
+
+		builder.errorMsg = "installplan 'nsname' cannot be empty"
+	}
+
+	return &builder
+}
+
+// GetInstallPlanBuilder returns installplan (first from the list).
+func GetInstallPlanBuilder(apiClient *clients.Settings, nsname string) (*Builder, error) {
```
Comment (kononovn) [inline pkg/olm/installplan.go:59]: Please re-name this function to ListInstallPlan and move it to the new file installplanlist.go.  This function has to return list > []*InstallPlanBuilder . You can use the first element of the returned list in your local test case code.

File: pkg/olm/installplan.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Builder provides a struct for installplan object from the cluster and a installplan definition.
+type Builder struct {
```
Comment (kononovn) [inline pkg/olm/installplan.go:17]: We have multiple elements under olm package. Please re-name Builder to InstallPlanBuilder

File: pkg/olm/installplan.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+		builder.errorMsg = "installplan 'nsname' cannot be empty"
+	}
+
+	return &builder
+}
+
+// GetInstallPlanBuilder returns installplan (first from the list).
+func GetInstallPlanBuilder(apiClient *clients.Settings, nsname string) (*Builder, error) {
+	if apiClient == nil {
```
Comment (kononovn) [inline pkg/olm/installplan.go:60]: Please remove this verification.

File: pkg/olm/installplan.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+	glog.V(100).Infof("Checking if installplan %s exists", builder.Definition.Name)
+
+	var err error
+	builder.Object, err = builder.apiClient.InstallPlans(builder.Definition.Namespace).Get(
+		context.Background(), builder.Definition.Name, metaV1.GetOptions{})
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
```
Comment (kononovn) [inline pkg/olm/installplan.go:107]: Please add Create and Update public function to this struct.

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix few comments above.

File: pkg/olm/installplan.go (hunk 3 above)
Comment (elenagerman) [reply pkg/olm/installplan.go:17]: yea, sorry it was developed as a separate pkg, and then we moved it ti the olm

File: pkg/olm/installplan.go (hunk 4 above)
Comment (elenagerman) [reply pkg/olm/installplan.go:60]: this method was already removed

File: pkg/olm/installplan.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	return &builder
+}
+
+// Create makes an InstallPlanBuilder in cluster and stores the created object in struct.
+func (builder *InstallPlanBuilder) Create() (*InstallPlanBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Creating the InstallPlanBuilder %s",
```
Comment (kononovn) [inline pkg/olm/installplan.go:64]: Please add namespace to the log message

File: pkg/olm/installplanlist.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ListInstallPlan returns installplan (first from the list).
```
Comment (kononovn) [inline pkg/olm/installplanlist.go:12]: Please edit function comment. This function returns just list of InstallPlanBuilder

File: pkg/olm/installplanlist.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+func ListInstallPlan(apiClient *clients.Settings, nsname string) ([]*InstallPlanBuilder, error) {
+	if apiClient == nil {
```
Comment (kononovn) [inline pkg/olm/installplanlist.go:14]: Please remove this verification.

**Review by kononovn — CHANGES_REQUESTED** [review]
Look very good, well done. Just 3 minor comments.

File: pkg/olm/installplanlist.go (hunk 7 above)
Comment (elenagerman) [reply pkg/olm/installplanlist.go:12]: oh yes, sorry, leftover from refactoring

File: pkg/olm/installplan.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	return &builder
+}
+
+// Create makes an InstallPlanBuilder in cluster and stores the created object in struct.
+func (builder *InstallPlanBuilder) Create() (*InstallPlanBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Creating the InstallPlanBuilder %v in namespace %v",
```
Comment (kononovn) [inline pkg/olm/installplan.go:64]: Please change it to Creating the InstallPlan

**Review by kononovn — DISMISSED** [review]
lgtm

File: pkg/olm/installplan.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+	return builder, err
+}
+
+// Exists checks whether the given installplan exists.
+func (builder *InstallPlanBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof("Checking if installplan %s exists", builder.Definition.Name)
```
Comment (trewest) [inline pkg/olm/installplan.go:82]: please add namespace to glog

File: pkg/olm/installplanlist.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ListInstallPlan returns list of the installplans found for specific namespace.
+func ListInstallPlan(apiClient *clients.Settings, nsname string) ([]*InstallPlanBuilder, error) {
+	installPlanList, err := apiClient.InstallPlans(nsname).List(context.Background(), v1.ListOptions{})
```
Comment (trewest) [inline pkg/olm/installplanlist.go:14]: Please check that nsname is not empty string

**Review by trewest — CHANGES_REQUESTED** [review]
Just a few comments

File: pkg/olm/installplan.go [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+	"github.com/openshift-kni/eco-goinfra/pkg/msg"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// InstallPlanBuilder provides a struct for installplan object from the cluster and a installplan definition.
```
Comment (achuzhoy) [inline pkg/olm/installplan.go:16]: and an installplan definition

File: pkg/olm/installplan.go [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+	return &builder
+}
+
+// Create makes an InstallPlanBuilder in cluster and stores the created object in struct.
+func (builder *InstallPlanBuilder) Create() (*InstallPlanBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Creating the InstallPlan %v in namespace %v",
```
Comment (achuzhoy) [inline pkg/olm/installplan.go:64]: let's use %s for consistency, i.e. 
"Creating InstallPlan %s in namespace %s"

File: pkg/olm/installplan.go [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+	glog.V(100).Infof("Checking if installplan %s exists", builder.Definition.Name)
+
+	var err error
+	builder.Object, err = builder.apiClient.InstallPlans(builder.Definition.Namespace).Get(
+		context.Background(), builder.Definition.Name, metaV1.GetOptions{})
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// Delete removes a installplan.
```
Comment (achuzhoy) [inline pkg/olm/installplan.go:91]: removes an installplan

File: pkg/olm/installplan.go [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+	return builder, err
+}
+
+// validate will check that the builder and builder definition are properly initialized before
+// accessing any member fields.
+func (builder *InstallPlanBuilder) validate() (bool, error) {
+	resourceCRD := "installplan"
+
+	if builder == nil {
+		glog.V(100).Infof("The builder %v is uninitialized", resourceCRD)
```
Comment (achuzhoy) [inline pkg/olm/installplan.go:138]: "The builder %s is uninitialized"

File: pkg/olm/installplan.go [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+
+		return false, fmt.Errorf("error: received nil %s builder", resourceCRD)
+	}
+
+	if builder.Definition == nil {
+		glog.V(100).Infof("The %v is undefined", resourceCRD)
```
Comment (achuzhoy) [inline pkg/olm/installplan.go:144]: "The %s is undefined"

File: pkg/olm/installplan.go [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+
+		builder.errorMsg = msg.UndefinedCrdObjectErrString(resourceCRD)
+	}
+
+	if builder.apiClient == nil {
+		glog.V(100).Infof("The builder %v apiclient is nil", resourceCRD)
```
Comment (achuzhoy) [inline pkg/olm/installplan.go:150]: %s

File: pkg/olm/installplan.go [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+
+		builder.errorMsg = fmt.Sprintf("%s builder cannot have nil apiClient", resourceCRD)
+	}
+
+	if builder.errorMsg != "" {
+		glog.V(100).Infof("The builder %v has error message: %w", resourceCRD, builder.errorMsg)
```
Comment (achuzhoy) [inline pkg/olm/installplan.go:156]: %s

File: pkg/olm/installplanlist.go [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+	if err != nil {
+		glog.V(100).Infof("Failed to list all installplan in namespace % due to %s",
```
Comment (achuzhoy) [inline pkg/olm/installplanlist.go:17]: "Failed to list all installplan in namespace %s due to %s"

File: pkg/olm/installplanlist.go [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ListInstallPlan returns list of the installplans found for specific namespace.
```
Comment (achuzhoy) [inline pkg/olm/installplanlist.go:12]: ListInstallPlan returns a list of installplans found under a specific namespace.

File: pkg/olm/installplan.go (hunk 15 above)
Comment (elenagerman) [reply pkg/olm/installplan.go:138]: for validate() method it's %v always was used. Are you sure I need to change?

Comment (achuzhoy) [reply pkg/olm/installplan.go]: Not sure. Just less indirection.
resourceCRD is a string  -just clearer.

**Review by trewest — APPROVED** [review]
/lgtm

**Review by kononovn — APPROVED** [review]
/lgtm

**Pull request merged**